packages/flutter/lib/src/foundation/assertions.dart

```diff
@@ -257,51 +257,103 @@ class FlutterErrorDetails {
 
+/// Helper class used for collecting different pieces of information
+/// for constructing an instance of [FlutterError]. It provides a number
+/// of methods with names starting with 'add' to make it
+/// convenient to add different parts to an error report. 
+/// In general, error parts will be displayed in the same order 
+/// of getting them added to the error builder. 
+/// {@tool sample}
+///
+/// Here is a stateful [Foo] widget. Its [State] uses the
+/// [SingleTickerProviderStateMixin] to implement the necessary
+/// [TickerProvider], creating its controller in the [initState] method and
+/// disposing of it in the [dispose] method. The duration of the controller is
+/// configured from a property in the [Foo] widget; as that changes, the
+/// [didUpdateWidget] method is used to update the controller.
+///
+/// ```dart
+/// throw FlutterErrorBuilder()
+///   ..addError('A short summary of the error. This is usually required.`)
+///   ..addDescription('A more detailed description of the error.')
+///   ..addFix('A resolution that is straightforward to implement.')
+///   ..addHint('A suggestion about a potential approach to resolving the error.')
+///   ..build(); // returns a FlutterError with all the specified parts.
+/// ```
+/// {@end-tool}
+/// {@tool sample}
+/// 
+/// See also:
+/// 
+/// * [WidgetErrorBuilder], which adds a few ready-made error elements 
+/// for reporting errors at the widget layer
+/// * [RenderErrorBuilder], which adds a few ready-made error elements 
+/// for reporting errors at the rendering layer
 class FlutterErrorBuilder {
+  /// Creates a [FlutterErrorBuilder]
   FlutterErrorBuilder() : _buildErrorCallback = null;
+  
+  /// Creates a [FlutterErrorBuilder] with its details computed only when needed.
+  /// Use if computing the error details may throw an exception or is expensive.
   FlutterErrorBuilder.lazy(this._buildErrorCallback);
 
+  /// Function called when building the error parts lazily. 
   final ErrorBuilderCallback<FlutterErrorBuilder> _buildErrorCallback;
 
   /// Overall error message.
   String error;
 
   final List<DiagnosticsNode> _parts = <DiagnosticsNode>[];
-
+  
+  /// Whether the error report is empty. 
   bool get isEmpty => _parts.isEmpty && error == null && _buildErrorCallback == null;
 
-  // Inject to add extra whitepsace between blocks of text.
+  /// Adds extra space between two parts of the error report.
+  /// The client can decide how the "separator" is rendered.
+  /// It falls back to an extra line break in the console.
   void addSeparator() {
     _parts.add(DiagnosticsNode.message(''));
   }
-
+  /// Adds a more elaborate description of the error.
+  /// It's strongly encouraged to show a summary of the error 
+  /// using [addError] before showing more details.
+  /// The description should include at least the following elements
+  /// * Claim: Explanation of the assertion failure or contract violation.
```

Comment:
This is interesting. We could have
addClaim, addGrounds, and addWaranty methods as well or instead of having addDescription?

Reply:
addDescription was just a name I added for the description part of an error message that didn't match any of the other categories.

Reply:
I'm hesitant to make the structure too rigid, especially a "warranty" can be as simple as glue words such as "because" or "due to". That said, addContract and addViolation seem to overlap with addDescription a bit. I'll examine some examples and see if we can consolidate some categories.

Reply:
sg